mongo-fixes/update-dimension-links/main.go

```diff
@@ -0,0 +1,189 @@
+	State             string                      `bson:"state,omitempty"                       json:"state,omitempty"`
+	Temporal          *[]models.TemporalFrequency `bson:"temporal,omitempty"                    json:"temporal,omitempty"`
+	TotalObservations *int                        `bson:"total_observations,omitempty"          json:"total_observations,omitempty"`
+	UniqueTimestamp   bson.MongoTimestamp         `bson:"unique_timestamp"                      json:"-"`
+	Version           int                         `bson:"version,omitempty"                     json:"version,omitempty"`
+}
+
+func main() {
+	flag.StringVar(&mongoURL, "mongo-url", mongoURL, "mongoDB URL")
+	flag.Parse()
+
+	ctx := context.Background()
+
+	if mongoURL == "" {
+		log.Event(ctx, "missing mongo-url flag", log.ERROR)
+		return
+	}
+
+	session, err := mgo.Dial(mongoURL)
+	if err != nil {
+		log.Event(ctx, "unable to create mongo session", log.Error(err), log.ERROR)
+		return
+	}
+	defer session.Close()
+
+	session.SetBatch(10000)
+	session.SetPrefetch(0.25)
+
+	log.Event(ctx, "successfully connected to mongo", log.INFO)
+
+	// Get all instances IDs
+	instanceIDs, err := getInstanceIDs(ctx, session)
+	if err != nil {
+		log.Event(ctx, "failed to get all instances", log.Error(err), log.ERROR)
+		return
+	}
+	log.Event(ctx, "successfully retrieved all instance ids", log.INFO)
+
+	// Create a backup collection
+	// dateTime formatted in YYMMDD_HHMMSS
+	dateTime := time.Now().Format("20060102_150405")
+	backupProgressBar := progressbar.Default(int64(len(instanceIDs)), "backup instance")
+	for _, id := range instanceIDs {
+		if err = addInstanceToBackup(ctx, session, id.ID, dateTime); err != nil {
+			log.Event(ctx, "failed to backup instances", log.Error(err), log.ERROR)
+			return
+		}
+		backupProgressBar.Add(1)
+		time.Sleep(40 * time.Millisecond)
+	}
+
+	log.Event(ctx, "successfully backed up all instance documents", log.INFO)
+
+	errorCount := 0
+	updateProgressBar := progressbar.Default(int64(len(instanceIDs)), "updating instance")
+
+	// loop over instances
+	for _, id := range instanceIDs {
+
+		// update the whole instance document
+		if err := updateInstance(ctx, session, id.ID); err != nil {
+			log.Event(ctx, "failed to update dimension of instance", log.Error(err), log.Data{"current_instance_id": id}, log.ERROR)
+			errorCount++
+		}
+
+		updateProgressBar.Add(1)
+		time.Sleep(40 * time.Millisecond)
+
+	}
+
+	if errorCount > 0 {
+		log.Event(ctx, "failed to update dimension of all instances", log.Data{"no_unsuccessful_updates": errorCount}, log.INFO)
+	} else {
+		log.Event(ctx, "successfully updated all instance documents", log.INFO)
+	}
+
+}
+
+func getInstanceIDs(ctx context.Context, session *mgo.Session) (results []MongoID, err error) {
+	s := session.Copy()
+	defer s.Close()
+
+	err = s.DB("datasets").C("instances").Find(bson.M{}).Select(bson.M{"_id": 1}).All(&results)
```

Comment:
we could limit the `Find` to get only those with `/v1/` in them?
something like (untested!)
```suggestion
	err = s.DB("datasets").C("instances").Find(bson.M{"dimensions.0.href": bson.M{ "$regex": "/v1/"}}).Select(bson.M{"_id": 1}).All(&results)
```

Reply:
But this would mean that the instances with `/v1/` would only be backed up as the `ids` are used to add instances to the backup

Reply:
The attraction of having a full backup, which is quick and easy to restore, (I think) hides a potential for data loss.
The database is constantly-changing, so any backup is a snapshot, and soon out-of-date.
To restore an out-of-date backup over the original collection would lose (and mask) any recent changes that had occurred in that collection.

Pros:
- smaller backup - only affected documents changed
- quicker backup
- more confidence of data integrity

Cons:
- restoring any documents is time-consuming (and even manual), because a comparison is needed (with the original) to ensure other parts of the doc have not changed